collections/src/main/scala/strawman/collection/mutable/ArrayBuilder.scala

```diff
@@ -0,0 +1,665 @@
+package strawman.collection
+package mutable
+
+import scala.{Array, Serializable, Byte, Short, Char, Int, Long, Float, Double, Boolean, Unit, AnyRef, Any}
+import scala.Predef.implicitly
+
+import scala.reflect.ClassTag
+
+/** A builder class for arrays.
+ *
+ *  @since 2.8
+ *
+ *  @tparam T    the type of the elements for the builder.
+ */
+abstract class ArrayBuilder[T] extends ReusableBuilder[T, Array[T]] with Serializable
+
+/** A companion object for array builders.
+ *
+ *  @since 2.8
+ */
+object ArrayBuilder {
+
+  /** Creates a new arraybuilder of type `T`.
+   *
+   *  @tparam T     type of the elements for the array builder, with a `ClassTag` context bound.
+   *  @return       a new empty array builder.
+   */
+  def make[T: ClassTag](): ArrayBuilder[T] = {
+    val tag = implicitly[ClassTag[T]]
+    tag.runtimeClass match {
+      case java.lang.Byte.TYPE      => new ArrayBuilder.ofByte().asInstanceOf[ArrayBuilder[T]]
+      case java.lang.Short.TYPE     => new ArrayBuilder.ofShort().asInstanceOf[ArrayBuilder[T]]
+      case java.lang.Character.TYPE => new ArrayBuilder.ofChar().asInstanceOf[ArrayBuilder[T]]
+      case java.lang.Integer.TYPE   => new ArrayBuilder.ofInt().asInstanceOf[ArrayBuilder[T]]
+      case java.lang.Long.TYPE      => new ArrayBuilder.ofLong().asInstanceOf[ArrayBuilder[T]]
+      case java.lang.Float.TYPE     => new ArrayBuilder.ofFloat().asInstanceOf[ArrayBuilder[T]]
+      case java.lang.Double.TYPE    => new ArrayBuilder.ofDouble().asInstanceOf[ArrayBuilder[T]]
+      case java.lang.Boolean.TYPE   => new ArrayBuilder.ofBoolean().asInstanceOf[ArrayBuilder[T]]
+      case java.lang.Void.TYPE      => new ArrayBuilder.ofUnit().asInstanceOf[ArrayBuilder[T]]
+      case _                        => new ArrayBuilder.ofRef[T with AnyRef]()(tag.asInstanceOf[ClassTag[T with AnyRef]]).asInstanceOf[ArrayBuilder[T]]
+    }
+  }
+
+  /** A class for array builders for arrays of reference types.
+   *
+   *  This builder can be reused.
+   *
+   *  @tparam T     type of elements for the array builder, subtype of `AnyRef` with a `ClassTag` context bound.
+   */
+  final class ofRef[T <: AnyRef : ClassTag] extends ArrayBuilder[T] {
```

Comment:
Can we make those `private`? Scala.js does not like them. We have an entirely different implementation of `ArrayBuilder` in Scala.js. For example see https://github.com/scala-js/scala-js/blob/7ea476aa62349f681bb2dbba16b7ffa6470bd786/scalalib/overrides-2.13/scala/collection/mutable/ArrayBuilder.scala#L32-L73

Reply:
These types are exposed as part of the API so that you can use them directly with unboxed primitive element types in all cases where you don't need polymorphism over primitive and reference types. Having to code against a non-specialized superclass means you need to box and unbox every single element you append to the Builder.